app/celery/nightly_tasks.py

```diff
@@ -334,3 +335,42 @@ def save_daily_notification_processing_time(bst_date=None):
             messages_within_10_secs=result.messages_within_10_secs,
         )
     )
+
+
+@notify_celery.task(name="delete-oldest-quarter-of-unneeded-notification-history")
+def delete_oldest_quarter_of_unneeded_notification_history():
+    # This retention limit is hardcoded for the moment. It is picked from
+    # https://github.com/alphagov/notifications-aws/blob/main/decisions/2022-12-01-notification-history-retention-period.md
+    # and can be increased in the future when we move 3 months into the next financial year
+    retention_limit = datetime(2022, 4, 1)
+
+    oldest_record = db.session.query(func.min(NotificationHistory.created_at)).one()[0]
```

Comment:
I think this query will read the entire notification history table because there's no index available on just `created_at`.

```
rdsbroker_6093d337_6918_4b97_9709_97529114eb90> explain select min(created_at) from notification_history;
+----------------------------------------------------------------------------------------------------------------+
| QUERY PLAN                                                                                                     |
|----------------------------------------------------------------------------------------------------------------|
| Finalize Aggregate  (cost=244986027.97..244986027.98 rows=1 width=8)                                           |
|   ->  Gather  (cost=244986027.75..244986027.96 rows=2 width=8)                                                 |
|         Workers Planned: 2                                                                                     |
|         ->  Partial Aggregate  (cost=244985027.75..244985027.76 rows=1 width=8)                                |
|               ->  Parallel Seq Scan on notification_history  (cost=0.00..237599580.80 rows=2954178780 width=8) |
+----------------------------------------------------------------------------------------------------------------+
EXPLAIN 5
Time: 0.061s
```

Given that, maybe just doing the delete query is fine? Otherwise we're scanning the table twice.

Alternatively let's put an index on `created_at`. I would have found one very useful for separate reasons over the last week.

---

But obviously we're doing this query to get the date of the oldest record so that we can work out what quarter of data to delete. An alternative to this would be to hard-code a start date (as you have below with 2019-10-01) and store/retrieve that date from redis in each run, adding 3 months each time. If the key isn't present, use 2019-10-01, otherwise whatever the value is (so that the first run works, but also so that if we lose redis, we recover - given a week or so).

Your call as to whether this optimisation is worth saving the double query. I think it could be, unless scanning notification history doesn't actually take as long as I fear it might.

Reply:
```
rdsbroker_6093d337_6918_4b97_9709_97529114eb90> explain select min(created_at) from notification_history;
+---------------------------------------------------------------------------------------------------------------------------->
| QUERY PLAN                                                                                                                 >
|---------------------------------------------------------------------------------------------------------------------------->
| Result  (cost=0.61..0.62 rows=1 width=8)                                                                                   >
|   InitPlan 1 (returns $0)                                                                                                  >
|     ->  Limit  (cost=0.58..0.61 rows=1 width=8)                                                                            >
|           ->  Index Only Scan using ix_notification_history_created_at on notification_history  (cost=0.58..204070599.34 ro>
|                 Index Cond: (created_at IS NOT NULL)                                                                       >
+---------------------------------------------------------------------------------------------------------------------------->
EXPLAIN 5
Time: 0.046s
rdsbroker_6093d337_6918_4b97_9709_97529114eb90> select min(created_at) from notification_history;
+----------------------------+
| min                        |
|----------------------------|
| 2019-06-20 00:00:06.372805 |
+----------------------------+
SELECT 1
Time: 0.039s
```

🎉